gui/app.py

```diff
@@ -20,32 +20,111 @@ class VideoThread(QThread):
+        # Run the play/pausable video 
+        while self._thread_running_flag:
+            # Send frames while the video is playing
+            while self._thread_running_flag and self._video_playing_flag:
+                self.__emit_frames()
+                self.msleep(int(1000 / 30))
+            
+            # Take a NOP while its paused
+            while self._thread_running_flag and not self._video_playing_flag:
+                self.msleep(10)
 
         # Shut down capturers
         for capture in self._captures:
             capture.release()
-
+    
+    def next_frame(self):
+        """Goes forward a frame if the video is paused"""
+        if not self._video_playing_flag:
+            prev_rewind_state = self._rewind
+            self._rewind = False
+            self.__emit_frames()
+            self._rewind = prev_rewind_state
+    
+    def prev_frame(self):
+        """Goes back a frame if the video is paused"""
+        if not self._video_playing_flag:
+            prev_rewind_state = self._rewind
+            self._rewind = True
+            self.__emit_frames()
+            self._rewind = prev_rewind_state
+
+    def toggle_rewind(self):
+        """Toggles the video rewind flag"""
+        self._rewind = not self._rewind
+
+    def toggle_play_pause(self):
+        """Toggles the video playing flag"""
+        self._video_playing_flag = not self._video_playing_flag
+    
     def stop(self):
-        """Sets run flag to False and waits for thread to finish"""
-        self._run_flag = False
+        """Sets the video playing & thread running flags to False and waits for thread to end"""
+        self._video_playing_flag = False
+        self._thread_running_flag = False
         self.wait()
 
+class VideoControlsWidget(QWidget):
+    def __init__(self, thread):
+        super().__init__()
+
+        self.thread = thread
+
+        self.horizontal_layout = QHBoxLayout(self)
+
+        self.play_pause_button = QPushButton(self)
+        self.play_pause_button.setText("Play/Pause")
+        self.play_pause_button.clicked.connect(self.thread.toggle_play_pause)
+        self.horizontal_layout.addWidget(self.play_pause_button)
+
+        self.toggle_rewind_button = QPushButton(self)
+        self.toggle_rewind_button.setText("Toggle Rewind")
+        self.toggle_rewind_button.clicked.connect(self.thread.toggle_rewind)
+        self.horizontal_layout.addWidget(self.toggle_rewind_button)
+
+        self.prev_frame_button = QPushButton(self)
+        self.prev_frame_button.setText("<")
+        self.prev_frame_button.clicked.connect(self.thread.prev_frame)
+        self.horizontal_layout.addWidget(self.prev_frame_button)
+
+        self.next_frame_button = QPushButton(self)
+        self.next_frame_button.setText(">")
+        self.next_frame_button.clicked.connect(self.thread.next_frame)
+        self.horizontal_layout.addWidget(self.next_frame_button)
+
+        self.setLayout(self.horizontal_layout)
```

Comment:
Could you try moving this to a separate file so we don't clutter this file up too much?